### PR TITLE
ieee802154: migrate `netdev_ieee802154_dst_filter` to a common ieee802154

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description

This PR migrates `netdev_ieee802154_dst_filter` to `ieee802154_dst_filter`.
This not only makes it not dependent to netdev, but will help with Kconfig modelling since now `netdev_ieee802154` can be represented as a pure netdev layer. I will add a follow up PR that shows this in action.

`netdev_ieee802154_dst_filter` is marked as deprecated and will be removed after 2022.01.

### Testing procedure
Check that `examples/gnrc_networking` compiles for e.g `nrf52840dk` (with `USEMODULE=`nrf802154_netdev_legacy`) and confirm that pinging still works.

### Issues/PRs references
None so far

### CODE REVIEW

File: sys/include/net/ieee802154.h [hunk 1]
```diff
@@ -396,6 +396,24 @@ int ieee802154_get_src(const uint8_t *mhr, uint8_t *src, le_uint16_t *src_pan);
  */
 int ieee802154_get_dst(const uint8_t *mhr, uint8_t *dst, le_uint16_t *dst_pan);
 
+/**
+ * @brief  Check whether a frame pass the IEEE 802.15.4 frame filter.
+ *
+ * A frame passes the frame filter only if:
+ * - The PAN ID matches the PAN ID of the frame filter
```
Comment (jia200x) [inline sys/include/net/ieee802154.h:403]:
```suggestion
 * - The PAN ID matches the PAN ID of the frame filter or the broadcast PAN ID
```

File: sys/include/net/ieee802154.h [hunk 2]
```diff
@@ -396,6 +396,24 @@ int ieee802154_get_src(const uint8_t *mhr, uint8_t *src, le_uint16_t *src_pan);
...
+ *
+ * @param[in] mhr           MAC header (PSDU)
+ * @param[in] pan           PAN ID of the frame filter.
+ * @param[in] short_addr    Short Address of the frame filter.
+ * @param[in] ext_addr      Extended Address of the frame filter.
+ *
+ * @return 0            if frame passes the frame filter.
+ * @return 1            if frame doesn't pass the frame filter.
+ */
+int ieee802154_dst_filter(const uint8_t *mhr, uint16_t pan, void *short_addr, void *ext_addr);
```
Comment (benpicco) [inline sys/include/net/ieee802154.h:415]: Why not

```suggestion
int ieee802154_dst_filter(const uint8_t *mhr, uint16_t pan, const uint16_t *short_addr, const eui64_t *ext_addr);
```

**Comment by MrKevinWeiss** [comment]
I pushed the changes and added unit tests, therefore can no longer ack... Hi @benpicco :wave:

**Comment by jia200x** [comment]
squashed

**Comment by MrKevinWeiss** [comment]
Seems like some `-Werror=strict-aliasing` issue on the esp8266 I thought I fixed that but maybe you can take another look @jia200x

**Comment by jia200x** [comment]
The test is fine. I just changed the short address from uint16_t to network_uint16_t.

**Comment by MrKevinWeiss** [comment]
Please squash, murdock is happy now.

File: sys/include/net/ieee802154.h [hunk 3]
```diff
@@ -396,6 +396,25 @@ int ieee802154_get_src(const uint8_t *mhr, uint8_t *src, le_uint16_t *src_pan);
...
+ * @param[in] mhr           MAC header (PSDU)
+ * @param[in] pan           PAN ID of the frame filter.
+ * @param[in] short_addr    Short Address of the frame filter.
+ * @param[in] ext_addr      Extended Address of the frame filter.
+ *
+ * @return 0            if frame passes the frame filter.
+ * @return 1            if frame doesn't pass the frame filter.
+ */
+int ieee802154_dst_filter(const uint8_t *mhr, uint16_t pan,
+                          network_uint16_t short_addr, const eui64_t *ext_addr);
```
Comment (benpicco) [inline sys/include/net/ieee802154.h:416]: Just to clarify: `pan` is `uint16_t` but `short_addr` is `network_uint16_t`?

Comment (jia200x) [reply sys/include/net/ieee802154.h:416]: yes :(
Unfortunately this is the way how it's presented in many `ieee802154_xxx` functions.

I would actually migrate everything to little endian values. We could the use all `ieee802154_get_xxx` operations without any kind of memcpy / reverse. But we would need to adapt many modules.

Comment (MrKevinWeiss) [reply sys/include/net/ieee802154.h:416]: Lets keep it simple for now, the reversing issue is out of scope of this PR.

**Comment by jia200x** [comment]
done!

**Comment by MrKevinWeiss** [comment]
I just saw #16576, maybe we can get 2 birds with one stone?

**Comment by jia200x** [comment]
> I just saw #16576, maybe we can get 2 birds with one stone?

I think it's #11191 

Although they touch the same file, the mechanism has changed quite a lot since the SubMAC was introduced. The packet in practice is parsed before it gets to the frame filter. So, I think we should keep them separate.

**Comment by MrKevinWeiss** [comment]
copy pasta :facepalm:

**Comment by MrKevinWeiss** [comment]
Good good.

**Comment by MrKevinWeiss** [comment]
ping @benpicco

**Review by benpicco — APPROVED** [review]
Communication using short addresses still works between `samr21-xpro` and `nrf52840dk`

**Pull request merged**